sklearn/decomposition/cdnmf_fast.pyx

```diff
@@ -0,0 +1,133 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+
+# Author: Mathieu Blondel, Tom Dupre la Tour for Cython version
+# License: BSD 3 clause
+
+import numpy as np
+cimport cython
+cimport numpy as np
+from libc.math cimport fabs, fmax, fmin
+
+
+def _update_cdnmf_fast(np.ndarray[double, ndim=2, mode='c'] W,
+                       np.ndarray[double, ndim=2, mode='c'] HHt,
+                       np.ndarray[double, ndim=2, mode='c'] XHt,
+                       int transpose_XHt):
+
+    cdef double violation = 0
+    cdef int n_components = W.shape[1]
+    cdef int n_length = W.shape[0]
+    cdef double grad
+    cdef double pg
+    cdef double hess
+    cdef int i
+    cdef int t
+    cdef int r
+
+    with nogil:
+        for i in range(n_length):
+            for t in range(n_components):
+
+                # gradient
+                # grad = GH[t, i] where GH = np.dot(HHt, H) - XHt
+                if transpose_XHt:
+                    grad = - XHt[t, i]
+                else:
+                    grad = - XHt[i, t]
+
+                for r in range(n_components):
+                    grad += HHt[t, r] * W[i, r]
+
+                # projected gradient
+                pg = fmin(0, grad) if W[i, t] == 0 else grad
+                violation += fabs(pg)
+
+                # Hessian
+                hess = HHt[t, t]
+
+                if hess != 0:
+                    W[i, t] = fmax(W[i, t] - grad / hess, 0)
+                
+    return violation
+
+
+def _update_greedy_cdnmf_fast(np.ndarray[double, ndim=2, mode='c'] W,
+                              np.ndarray[double, ndim=2, mode='c'] G,
+                              np.ndarray[double, ndim=2, mode='c'] HHt,
+                              int max_inner, double tol):
+    cdef double violation = 0.
+    cdef double pg
+    cdef int n_length = W.shape[0]
```

Comment:
W is `(n_samples, n_components)`
H.T is `(n_features, n_components)`
I changed the name to be more clear:
`cdef int n_samples = W.shape[0]  # n_features for H update`